File: web/services/storage/reindex.ts
import { defineStore } from 'pinia'
import { ref, computed } from 'vue'

import Api from '../api'
import * as cryptfns from '../cryptfns'
import * as meta from './meta'
import { downloadAndDecrypt } from './download/sync'

import type { AppFile, EncryptedAppFile, KeyPair } from 'types'

/**
 * Files handled per round. Small enough that cancelling feels immediate and
 * that a note download never blocks the bar for long, large enough that the
 * request overhead is not what dominates the sweep.
 */
const BATCH_SIZE = 10

export interface ReindexRequest {
  name_hash: string
  /** Account fingerprint the tags were keyed under. The server rejects a
   *  value that is not the live `users.fingerprint`, so a sweep that started
   *  under a discarded key cannot retire files from the pending list. */
  fingerprint: string
  search_tokens_root: string[]
  search_tokens_file: string[]
  /** Content digests re-keyed under the file's search key, replacing the
   *  bare digests migrated rows still carry. */
  md5?: string
  sha1?: string
  sha256?: string
  blake2b?: string
  /** The tags that make those digests findable in search. Separate from the
   *  word tokens because they land in the digest scopes, which renames never
   *  touch. */
  digest_tokens_root?: string[]
  digest_tokens_file?: string[]
}

/** Bare digest lengths in hex, per algorithm. A keyed tag is 32 hex chars. */
const BARE_DIGEST_LENGTH = { md5: 32, sha1: 40, sha256: 64, blake2b: 128 } as const

type DigestName = keyof typeof BARE_DIGEST_LENGTH

/**
 * The digest columns a sweep may re-key: the ones still holding a bare
 * digest. A file can pass through the sweep twice — a note edited before the
 * sweep already got a keyed sha256 from its save, and a failed
 * crypto-migration ceremony can run the whole sweep once on the old key —
 * and keying a keyed value corrupts the column beyond repair. Shape decides
 * for sha1/sha256/blake2b; bare MD5 is 32 hex chars like a tag, so it goes
 * by its siblings: every writer that stored an MD5 stored a SHA-256 next to
 * it, so a row keying any sibling is a bare row and one keying none is
 * already done.
 */
export function bareDigests(
  file: Partial<Record<DigestName, string>>
): Partial<Record<DigestName, string>> {
  const bare = (name: DigestName): string | undefined => {
    const digest = file[name]
    if (!digest || digest.length !== BARE_DIGEST_LENGTH[name]) return undefined
    return /^[0-9a-f]+$/i.test(digest) ? digest : undefined
  }

  const digests: Partial<Record<DigestName, string>> = {
    sha1: bare('sha1'),
    sha256: bare('sha256'),
    blake2b: bare('blake2b')
  }
  if (digests.sha1 || digests.sha256 || digests.blake2b) {
    digests.md5 = bare('md5')
  }

  return digests
}

/**
 * Rebuild the search index for files that predate keyed tags.
 *
 * The re-key migration dropped every old index row, and nothing server-side
 * can rebuild them — the tags are keyed on material only this client holds. So
 * each client walks its own files once.
 *
 * Progress needs no bookkeeping of its own: the server reports a file as
 * pending exactly while its `name_hash` is the blank the migration left, so
 * the keyed hash every re-index writes is what marks it done. Closing the
 * tab, cancelling, or losing the connection all resume from the same place,
 * which is simply "whatever is still pending".
 */
export const store = defineStore('reindex', () => {
  const total = ref(0)
  const done = ref(0)
  const running = ref(false)
  const background = ref(false)
  const failed = ref(0)

  /**
   * Stop after the batch in flight. Whatever is left is still pending
   * server-side, so the next session picks it up from there.
   */
  const cancelled = ref(false)

  const progress = computed(() => (total.value === 0 ? 0 : Math.round((done.value / total.value) * 100)))
  const visible = computed(() => running.value && !background.value && !cancelled.value)

  async function fetchPending(): Promise<EncryptedAppFile[]> {
    const response = await Api.get<EncryptedAppFile[]>('/api/storage/reindex')

    return response?.body ?? []
  }

  /**
   * How many files still need doing. Cheap enough to call on unlock to decide
   * whether the modal is worth showing at all.
   */
  async function countPending(): Promise<number> {
    return (await fetchPending()).length
  }

  /**
   * A note's body is indexed word for word alongside its name, which is why
   * the old scheme leaked note contents and not just names. Rebuilding that
   * means fetching and decrypting the note — there is no shortcut, the server
   * holds only ciphertext — and the name rides along so a swept note carries
   * the same tokens a saved one does.
   */
  async function textFor(file: AppFile): Promise<string> {
    const name = file.name ?? ''
    if (!file.editable || !file.key) {
      return name
    }

    const bytes = await downloadAndDecrypt(file)

    return `${name}\n${new TextDecoder().decode(bytes)}`
  }

  async function reindexOne(keypair: KeyPair, encrypted: EncryptedAppFile): Promise<void> {
    const privateKey = keypair.wrappingPrivate || keypair.input

    if (!privateKey) {
      throw new Error('Cannot re-index without an unlocked private key')
    }

    const file = { ...encrypted, ...(await meta.decrypt(encrypted, privateKey)) } as AppFile

    if (!file.key || !file.name) {
      throw new Error('file has no key or name')
    }

    const fingerprint = keypair.fingerprint
    if (!fingerprint) {
      throw new Error('Cannot re-index without the account fingerprint')
    }

    const rootKey = cryptfns.searchRootKey(keypair)
    const fileKey = cryptfns.searchFileKey(file.key)
    const indexed = await textFor(file)

    const body: ReindexRequest = {
      name_hash: cryptfns.searchTag(rootKey, file.name),
      fingerprint,
      search_tokens_root: cryptfns.searchTags(rootKey, indexed),
      search_tokens_file: cryptfns.searchTags(fileKey, indexed)
    }

    // Migrated rows still carry bare content digests — the third copy of the
    // same leak. Re-key each from the stored value (no re-download needed)
    // and index it in both scopes, which is what makes pasting a digest into
    // search find the file. Only values still in the bare shape are re-keyed;
    // `bareDigests` above says why.
    const digests = bareDigests(file)

    for (const name of ['md5', 'sha1', 'sha256', 'blake2b'] as const) {
      const digest = digests[name]
      if (!digest) continue
      body[name] = cryptfns.searchTag(fileKey, digest)
      body.digest_tokens_root = body.digest_tokens_root ?? []
      body.digest_tokens_file = body.digest_tokens_file ?? []
      body.digest_tokens_root.push(`${cryptfns.searchTag(rootKey, digest)}:1`)
      body.digest_tokens_file.push(`${cryptfns.searchTag(fileKey, digest)}:1`)
    }

    await Api.put<ReindexRequest, AppFile>(`/api/storage/${file.id}/reindex`, undefined, body)
  }

  /**
   * Walk every pending file in batches until the server reports none left.
   *
   * A file that throws is counted and skipped rather than aborting the sweep —
   * one unreadable file should not cost the user their whole index. It stays
   * pending, so the next run tries it again.
   */
  async function run(keypair: KeyPair): Promise<void> {
    if (running.value) return

    running.value = true
    cancelled.value = false
    background.value = false
    done.value = 0
    failed.value = 0

    // By id rather than by count, and at most one attempt per file per run.
    // The server's answer to an attempt is authoritative either way: success
    // takes the file off the pending list, and a failure leaves it there for
    // the next run. Re-attempting inside this run would change nothing the
    // second time — and if a server ever kept reporting a successfully
    // re-indexed file as pending, retrying it would spin this loop forever.
    const failedIds = new Set<string>()
    const attemptedIds = new Set<string>()
    const seenIds = new Set<string>()

    try {
      let pending = await fetchPending()
      pending.forEach((f) => seenIds.add(f.id))
      total.value = seenIds.size

      while (pending.length > 0) {
        for (let i = 0; i < pending.length; i += BATCH_SIZE) {
          if (cancelled.value) return

          const batch = pending.slice(i, i + BATCH_SIZE)

          await Promise.all(
            batch.map(async (encrypted) => {
              try {
                await reindexOne(keypair, encrypted)
              } catch {
                failedIds.add(encrypted.id)
              } finally {
                attemptedIds.add(encrypted.id)
                done.value += 1
                failed.value = failedIds.size
              }
            })
          )
        }

        if (cancelled.value) return

        // The server pages at 500, so keep asking until nothing fresh comes
        // back. Stopping when the next page is no smaller than this one broke
        // after a single page on any account past that limit: page two is also
        // full, so it read as "no progress" and the sweep quit with most of
        // the account still unindexed. Only what this run has not touched
        // counts as fresh.
        const next = (await fetchPending()).filter((f) => !attemptedIds.has(f.id))
        if (!next.length) break

        next.forEach((f) => seenIds.add(f.id))
        total.value = seenIds.size
        pending = next
      }
    } finally {
      running.value = false
    }
  }

  /** Dismiss the modal and stop. Whatever is left stays pending for next time. */
  function cancel() {
    cancelled.value = true
  }

  /** Dismiss the modal but let the sweep finish. */
  function continueInBackground() {
    background.value = true
  }

  return {
    total,
    done,
    failed,
    running,
    background,
    cancelled,
    progress,
    visible,
    countPending,
    run,
    cancel,
    continueInBackground
  }
})

File: storage/src/data/reindex.rs
//! Re-index one file against the keyed search scheme.
//!
//! The migration that introduced keyed tags dropped every old index row,
//! because those rows were reversible and keeping them through a transition
//! would have kept the readable copy alive indefinitely. Nothing server-side
//! can rebuild them: the tags are keyed on material only the client holds. So
//! each client walks its own files once and re-indexes them through here.

use ::error::AppResult;
use entity::file_tokens::{DigestTags, SearchTags};
use serde::{Deserialize, Serialize};
use validr::*;

#[derive(Clone, Debug, Serialize, Deserialize, Default)]
pub struct Reindex {
    /// Re-keyed `name_hash`. Migrated rows still carry the old unsalted
    /// digest of the plaintext name, which is the same leak in a second
    /// place, so the sweep replaces it alongside the tags.
    pub name_hash: Option<String>,
    /// The account fingerprint the tags were keyed under. The root search
    /// key is derived from the private key, so a session that started the
    /// sweep before a key-rotation ceremony committed would otherwise write
    /// a 32-hex `name_hash` that looks "done" and never matches a query
    /// under the new key. Compared to `users.fingerprint` at write time;
    /// a mismatch leaves the file pending.
    pub fingerprint: Option<String>,
    pub search_tokens_root: Option<Vec<String>>,
    pub search_tokens_file: Option<Vec<String>>,
    /// Note-body tokens, written to `source=content`. Absent on regular files
    /// and on clients that still mix name and body into `search_tokens_*`.
    pub content_tokens_root: Option<Vec<String>>,
    pub content_tokens_file: Option<Vec<String>>,
    /// Content digests re-keyed under the file's search key, replacing the
    /// bare digests migrated rows still carry — the third copy of the same
    /// leak. The sweep computes them from the stored values, so nothing has
    /// to be re-downloaded, and a client that can hold the file key can
    /// still run its resume equality check against the keyed form.
    pub md5: Option<String>,
    pub sha1: Option<String>,
    pub sha256: Option<String>,
    pub blake2b: Option<String>,
    /// The digest tags that make the re-keyed values findable by pasting a
    /// digest into search. Separate from the word tokens because they land in
    /// the digest scopes, which renames never touch.
    pub digest_tokens_root: Option<Vec<String>>,
    pub digest_tokens_file: Option<Vec<String>>,
}

/// The keyed digest columns a re-index rewrites, `None` meaning "leave it".
pub struct KeyedHashes {
    pub md5: Option<String>,
    pub sha1: Option<String>,
    pub sha256: Option<String>,
    pub blake2b: Option<String>,
}

impl Validation for Reindex {
    fn rules(&self) -> Vec<Rule<Self>> {
        vec![rule_required!(name_hash), rule_required!(fingerprint)]
    }
}

impl Reindex {
    pub fn into_parts(
        self,
    ) -> AppResult<(
        String,
        String,
        SearchTags,
        SearchTags,
        DigestTags,
        KeyedHashes,
    )> {
        let data = self.validate()?;
        let name_hash = data.name_hash.unwrap();
        let fingerprint = data.fingerprint.unwrap();

        // This route exists to replace the reversible digest, so of all
        // places it must not accept one back. Same refusal as create and
        // rename: a keyed hash is half the length, so the shapes never
        // collide.
        if cryptfns::search::is_legacy_name_hash(&name_hash) {
            return Err(::error::Error::UpgradeRequired(
                "client_too_old_for_search".to_string(),
            ));
        }

        // The same refusal for the digest columns themselves: a sweep that
        // would write a bare digest back defeats its own purpose. MD5 shares
        // the tag's shape and cannot be told apart here.
        for value in [&data.md5, &data.sha1, &data.sha256, &data.blake2b]
            .into_iter()
            .flatten()
        {
            if cryptfns::search::is_bare_digest(value) {
                return Err(::error::Error::UpgradeRequired(
                    "client_too_old_for_search".to_string(),
                ));
            }
        }

        Ok((
            name_hash,
            fingerprint,
            SearchTags::new(data.search_tokens_root, data.search_tokens_file),
            SearchTags::new(data.content_tokens_root, data.content_tokens_file),
            DigestTags::new(data.digest_tokens_root, data.digest_tokens_file),
            KeyedHashes {
                md5: data.md5,
                sha1: data.sha1,
                sha256: data.sha256,
                blake2b: data.blake2b,
            },
        ))
    }
}

#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn name_hash_is_required() {
        assert!(Reindex::default().into_parts().is_err());
    }

    #[test]
    fn fingerprint_is_required() {
        let result = Reindex {
            name_hash: Some("a".repeat(32)),
            ..Default::default()
        }
        .into_parts();

        assert!(result.is_err());
    }

    #[test]
    fn a_legacy_digest_is_refused() {
        let result = Reindex {
            name_hash: Some(cryptfns::sha256::digest("secret name".as_bytes())),
            fingerprint: Some("fp".to_string()),
            ..Default::default()
        }
        .into_parts();

        assert!(
            matches!(result, Err(::error::Error::UpgradeRequired(_))),
            "the route built to replace the reversible digest accepted one back"
        );
    }

    #[test]
    fn a_bare_digest_column_is_refused() {
        let result = Reindex {
            name_hash: Some("a".repeat(32)),
            fingerprint: Some("fp".to_string()),
            sha1: Some(cryptfns::sha256::digest("bytes".as_bytes())),
            ..Default::default()
        }
        .into_parts();

        assert!(
            matches!(result, Err(::error::Error::UpgradeRequired(_))),
            "a digest column accepted a value that is not a keyed tag"
        );
    }

    #[test]
    fn tags_and_keyed_hashes_ride_along_with_the_name_hash() {
        let (name_hash, fingerprint, tags, content, digests, hashes) = Reindex {
            name_hash: Some("abc".to_string()),
            fingerprint: Some("fp".to_string()),
            search_tokens_root: Some(vec!["a3f1:2".to_string()]),
            search_tokens_file: Some(vec!["9c22:1".to_string()]),
            sha256: Some("b".repeat(32)),
            digest_tokens_root: Some(vec![format!("{}:1", "c".repeat(32))]),
            ..Default::default()
        }
        .into_parts()
        .unwrap();

        assert_eq!(name_hash, "abc");
        assert_eq!(fingerprint, "fp");
        assert_eq!(tags.root.unwrap().len(), 1);
        assert_eq!(tags.file.unwrap().len(), 1);
        assert!(content.root.is_none());
        assert!(content.file.is_none());
        assert_eq!(digests.root.unwrap().len(), 1);
        assert!(digests.file.is_none());
        assert_eq!(hashes.sha256.unwrap(), "b".repeat(32));
        assert!(hashes.md5.is_none());
    }

    #[test]
    fn content_tags_ride_along_separately_from_the_name() {
        let (_, _, tags, content, _, _) = Reindex {
            name_hash: Some("abc".to_string()),
            fingerprint: Some("fp".to_string()),
            search_tokens_root: Some(vec!["a3f1:2".to_string()]),
            content_tokens_root: Some(vec!["b4e2:1".to_string()]),
            content_tokens_file: Some(vec!["c5d3:1".to_string()]),
            ..Default::default()
        }
        .into_parts()
        .unwrap();

        assert_eq!(tags.root.unwrap().len(), 1);
        assert!(tags.file.is_none());
        assert_eq!(content.root.unwrap().len(), 1);
        assert_eq!(content.file.unwrap().len(), 1);
    }
}
